### PR TITLE
Avoid intermediate NaN in `multiply_no_nan`.

### PR DESCRIPTION
**Pull request by rootkiller6788**
## Problem

`math.multiply_no_nan(x, y)` is meant to return 0 when `y == 0`, even if `x` is NaN or infinite. The previous implementation computed

```python
jnp.where(y == 0, jnp.zeros((), dtype=dtype), x * y)
```

which eagerly evaluates `x * y` for every element. When `x` is infinite and `y` is zero, `inf * 0` produces an intermediate NaN that `jnp.where` then masks out. The result is numerically correct, but the intermediate NaN is still observed by `checkify`'s NaN checks, so enabling `checkify.nan_checks` on code that calls `multiply_no_nan` (e.g. `Categorical.log_prob` when some probabilities are zero) raises a spurious `JaxRuntimeError` (see #187).

## Fix

Zero out `x` where `y == 0` *before* multiplying, so `0 * y` is computed instead of `x * 0`:

```python
jnp.where(y == 0, jnp.zeros((), dtype=dtype), x) * y
```

This produces identical results and gradients (the custom JVP is unchanged), but never materializes the NaN.

## Test

Added `test_multiply_no_nan_checkify`, which wraps `multiply_no_nan(-inf, 0)` in `checkify.checkify(..., errors=checkify.nan_checks)` and asserts no NaN check is triggered. This test fails on the previous implementation and passes after the fix. Existing `test_multiply_no_nan` and `test_multiply_no_nan_grads` still pass.

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
The primal path is now `checkify`-safe, but the custom JVP can still materialize the same `0 * inf` NaN. `scale_maybe_symbolic` only special-cases `SymbolicZero`; with a concrete zero tangent for `y`, `tangent_out_2` evaluates `0 * x`, so e.g. differentiating at `x=-inf, y=0` produces a NaN tangent and `checkify.nan_checks` still fires. Please make this tangent product no-NaN for a zero tangent as well and add a `checkify` + `jax.jvp` regression.

**Comment by rootkiller6788** [comment]
Thanks — you're right, and this is fixed now. tangent_out_2 scales by _zero_where_y_is_zero(x, y), the same guarded operand the primal uses, so it evaluates 0 * y rather than x * 0. A concrete zero tangent for y now yields a zero tangent instead of 0 * inf, and checkify + jax.jvp no longer fires; added test_multiply_no_nan_checkify_jvp. Worth flagging how this has to be done: branching on the tangent is not an option. scale_maybe_symbolic guards the SymbolicZero instance, but any elementwise tangent == 0 test makes the rule non-linear in the tangents, and reverse mode transposes it — fancy_bilinear_transpose asserts exactly one operand of a bilinear primitive is the transposed input, so jax.grad breaks for every input, not just x=-inf (I tried four other formulations; all of them broke jax.grad((3., 4.))). Hence the mask must come from primals. The consequence is a deliberate change to the gradient convention: ∂f/∂y at y == 0 is now 0 rather than x — that is exactly what makes the product NaN-free, and it matches what the symbolic-zero path already returned. test_multiply_no_nan_grads asserted the old (y, x) at y = 0 and now asserts (y, 0.). If you'd rather keep ∂f/∂y = x at y == 0, the zero-tangent NaN is not removable, and I'd rather fix the convention than leave checkify broken — but it's your call. One case I did not cover, since it's the mirror image and equally unfixable in-rule: jax.jvp(f, (-inf, 0.), (inf, 0.)) still gives a NaN from x_dot * y. Happy to file it separately if you think it's worth tracking. Verified with JAX 0.10.2: math_test.py 13 passed; bernoulli, categorical, multinomial, one_hot_categorical suites all green.

**Review by sylvesterkaczmarek — CHANGES_REQUESTED** [review]
The checkify/JVP NaN is fixed, but the current solution changes the finite gradient convention more broadly than necessary: for every finite x with y == 0, d/dy becomes 0 instead of the previous x.

Only non-finite x makes a concrete zero y-tangent produce 0 * inf/NaN. Could the JVP mask x only where y == 0 and x is non-finite, preserving the existing finite gradient while still making the -inf/0 zero-tangent case checkify-safe? A regression such as x=3, y=0 retaining grad_y=3 alongside the current -inf checkify case would pin both behaviors.
